src/Pelagos/Util/DOIutil.php

```diff
@@ -375,4 +334,51 @@ function ($matches) {
             $input
         );
     }
+
+    /**
+     * This function generates a new DOI string that is not already in use per datacite.
+     *
+     * @return string A unique, unused DOI string.
+     */
+    public function generateDoi()
+    {
+        $encoder = new Base32CrockfordEncoder();
+        // 1099511627775 encodes to the longest 8 character Crockford 32 string.
+        $max = 1099511627775;
+        // Start at 1 (0 is problematic as library does not produce checksum for 0.)
+        $random = mt_rand(1, $max);
+        // Add shoulder and remove the checksum character.
+        $doi = $this->doishoulder . '/' . substr($encoder->encode($random), 0, -1);
+        while ($this->checkDoiExistsExternal($doi)) {
+            $random = mt_rand(1, $max);
+            $doi = $this->doishoulder . '/' . substr($encoder->encode($random), 0, -1);
+        }
+        return $doi;
+    }
+
+    /**
+     * This function queries Datacite to determine if the DOI is already in use.
+     *
+     * @param string $doi The DOI to be checked.
+     *
+     * @throws HttpClientErrorException If a non 400 error is encountered.
+     *
+     * @return boolean True if returned if the DOI exists, false if it does not.
+     */
+    public function checkDoiExistsExternal(string $doi): bool
+    {
+        try {
+            $doiUtil = new DOIutil();
+            $doiUtil->getDOIMetadata($doi);
+        } catch (HttpClientErrorException $exception) {
+            if ($exception->getCode() == 400) {
+                // No, this DOI does not exist upstream (Datacite) thus can be used.
+                return false;
+            } else {
+                throw new HttpClientErrorException($exception->getMessage(), $exception->getCode());
```

Comment:
We are throwing this exception class inside the same Exception class that is being caught. And this thrown exception is not being caught anywhere

Reply:
It's re-throwing all other 400's except 400 itself. It will perculate up to DoiConsumer's issueDoi() method, which would result in the attempt being logged. 500's aren't caught at all, and would also perculate up to this point for a retry 10 minutes later. Previously mintDOI would have thrown the 500. It still gets handled the same way.